StrongInject.Generator/Visitors/DependencyCheckerVisitor.cs

```diff
@@ -249,17 +246,15 @@ public bool IsScopeAsync
             public HashSet<IParameterSymbol>? UsedParams { get; set; }
 
 
-            public bool AddCurrentlyVisiting(InstanceSource source) =>
-                source.Scope == Scope.SingleInstance
-                    ? CurrentlyVisitingSingletons.Add(source)
-                    : CurrentlyVisitingNonSingletons.Add(source);
-            
-            public bool RemoveCurrentlyVisiting(InstanceSource source) =>
-                source.Scope == Scope.SingleInstance
-                    ? CurrentlyVisitingSingletons.Remove(source)
-                    : CurrentlyVisitingNonSingletons.Remove(source);            
-            public HashSet<InstanceSource> CurrentlyVisitingSingletons { get; set; }
-            public HashSet<InstanceSource> CurrentlyVisitingNonSingletons { get; set; }
+            public bool AddCurrentlyVisiting(InstanceSource source, InstanceSourcesScope scope, out InstanceSourcesScope existingScope)
+            {
+                CurrentlyVisiting.AddOrUpdate(source, scope, out existingScope);
+                return existingScope is null;
+            }
+
+            public bool RemoveCurrentlyVisiting(InstanceSource source) => CurrentlyVisiting.Remove(source);
+
+            public Dictionary<InstanceSource, InstanceSourcesScope> CurrentlyVisiting { get; set; }
```

Comment:
Why is direct mutation from outside the class allowed as well as having rich updating methods `AddCurrentlyVisiting` and `RemoveCurrentlyVisiting`?